### DISCUSSION TITLE
Is it possible to use GitHub Actions to upload to bintray?

### DISCUSSION THREAD

**Discussion by Andre601** [community/community#25872] [Actions]
<p>I want to use GitHub Actions to build my java project and upload it to Bintray when creating a new release.</p>
<p>Now my question is: Is this even possible?</p>
<p>I use gradle and therefore attempted to use the gradle workflow.</p>
<p>The gradle.yml I setup looks like this:</p>
<pre><code>name: Java CI

on:
  release:
    branches:
    - master

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v1
    - name: Set up JDK 1.8
      uses: actions/setup-java@v1
      with:
        java-version: 1.8
    - name: Build with Gradle
      env:
        bintrayUser: ${{ secrets.bintrayUsername }}
        bintrayApiKey: ${{ secrets.bintrayApiKey }}
      run: ./gradlew bintrayUpload
</code></pre>
<p>The ${{ secrets.[…] }} are setup in the Repository’s Secrets setting.</p>
<p>I also use the following build.gradle:</p>
<pre><code>plugins{
    id 'com.jfrog.bintray' version '1.8.4'
    id 'java-library'
    id 'idea'
    id 'maven-publish'
    id 'maven'
}

group = 'com.andre601'
version = '2.3.1'

sourceCompatibility = 1.8

jar{
    baseName = 'JavaBotBlockAPI'
}

repositories{
    mavenCentral()
    jcenter()
}

dependencies{
    api group: 'com.squareup.okhttp3', name: 'okhttp', version: '3.11.0'
    api group: 'org.json', name: 'json', version: '20180813'
    api group: 'org.jetbrains', name: 'annotations', version: '16.0.2'
    api(group: 'net.dv8tion', name: 'JDA', version: '4.0.0_45'){
        exclude(module: 'opus-java')
    }
    api group: 'com.github.ben-manes.caffeine', name: 'caffeine', version: '2.7.0'
}

task sourcesJar(type: Jar, dependsOn: classes){
    classifier = 'sources'
    from sourceSets.main.allSource
}

task javadocJar(type: Jar, dependsOn: javadoc){
    classifier = 'javadoc'
    from javadoc.destinationDir
}

artifacts{
    archives sourcesJar
    archives javadocJar
}

bintray{
    user = System.getenv("bintrayUser")
    key = System.getenv("bintrayApiKey")
    publications = ['JavaBotBlockAPIRelease']
    pkg{
        repo = 'maven'
        name = 'JavaBotBlockAPI'
        licenses = ['MIT']
        vcsUrl = 'https://github.com/botblock/JavaBotBlockAPI'
        version{
            name = project.version
            released = new Date()
        }
    }
}

publishing{
    publications{
        JavaBotBlockAPIRelease(MavenPublication){
            from components.java
            groupId group
            artifactId archivesBaseName
            version version
            artifact javadocJar
            artifact sourcesJar
        }
    }
}
</code></pre>
<p>The issue now is that the Action fails at the “Build with Gradle” task with the error "./gradlew: Permission denied</p>
<p>Is there any way I could make this work and have GitHub automatically upload (deploy?) the new files to the bintray sites or would I need to search for an alternative?</p>

**Comment by chrispat** [comment]
<p>Could you try to chmod +x gradew before you run the script?  My guess is the execute bit is not set on that file in the repo.</p>

**Comment by Andre601** [comment]
<p>So by that you mean changing the gradle.yml to this?</p>
<pre><code>name: Java CI

on:
  release:
    branches:
    - master


jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v1
    - name: Set up JDK 1.8
      uses: actions/setup-java@v1
      with:
        java-version: 1.8
    - name: Grant rights
      run: chmod +x build.gradle
    - name: Build with Gradle
      env:
        bintrayUser: ${{ secrets.bintrayUsername }}
        bintrayApiKey: ${{ secrets.bintrayApiKey }}
      run: ./gradlew bintrayUpload
</code></pre>

**Comment by ethomson** [comment, marked as answer]
<p>No, it sounds like the gradlew script in your repository is not marked as executable.</p>
<p>If you’re on a Unix platform, run:</p>
<pre><code>chmod 0755 gradlew &amp;&amp; git add gradlew
</code></pre>
<p>If you’re on Windows, run:</p>
<pre><code>git update-index --chmod=+x gradlew
</code></pre>

**Comment by Andre601** [comment]
<p>I’m now a little bit confused…</p>
<p>I did what you said and it seemed to have partially worked.</p>
<p>I say partially because it both uploaded files (Even tho I’m not sure if it’s all and if they are uncorrupted/completed and all) to bintray (See first image) but also did throw an error (see second image).</p>
<p>Image 1:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/f5d08dd6309d42c4092b18940c99470df7993b2b.png" data-download-href="/uploads/short-url/z4zNVnIrBqLfA7bldBqJvKSERnB.png?dl=1" title="Bintray.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/11576465/181093783-c9a91180-cd97-4992-b814-6ab33d2d04aa.png" alt="Bintray.png" data-base62-sha1="z4zNVnIrBqLfA7bldBqJvKSERnB" width="690" height="251" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/f5d08dd6309d42c4092b18940c99470df7993b2b_2_690x251.png, https://aws1.discourse-cdn.com/github/optimized/2X/f/f5d08dd6309d42c4092b18940c99470df7993b2b_2_1035x376.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/f/f5d08dd6309d42c4092b18940c99470df7993b2b.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/f5d08dd6309d42c4092b18940c99470df7993b2b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Bintray.png</span><span class="informations">1132×413 22.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Image 2:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/59a50540b0d4c694194d3286ff38772cf94025c3.png" data-download-href="/uploads/short-url/cN2352ThHziuuXzvNhLcnND6w9R.png?dl=1" title="GitHub Actions error.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/11576465/181093786-0451ed10-3057-4bb2-9871-2bd1d3338615.png" alt="GitHub Actions error.png" data-base62-sha1="cN2352ThHziuuXzvNhLcnND6w9R" width="690" height="359" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/59a50540b0d4c694194d3286ff38772cf94025c3_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">GitHub Actions error.png</span><span class="informations">1043×543 27.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by oldskool** [comment]
<p>From the look of it you’re trying to push/tag a version that already exists?</p>
<p>The line saying “Could not create version 2.3.6” seems to indicate that version is conflicting. I’m not familiar with bintray nor gradle, but could it be that you already have a version 2.3.6 uploaded and it only accepts one upload per version?</p>

**Comment by Andre601** [comment]
<p>It does indeed only accept one release per version.</p>
<p>However, I never released a version 2.3.6 before so this shouldn’t be the problem… I’ll update the version and look again.</p>
<p>Update: Nope. It used a separate version and it didn’t work. Same error like before.</p>

**Comment by Andre601** [comment]
<p>Found the issue.</p>
<p>GitHub Actions triggered 3 times when I published a new release, causing the first that finished to upload the files (therefore creating the release on bintray) and the other two to fail because it’s already uploaded.</p>
<p>Adding the “types: [published]” check to the trigger solved this issue and the CI could finish without errors.</p>
